crates/worker/src/partition/state_machine/lifecycle/trim.rs

```diff
@@ -0,0 +1,472 @@
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::partition::state_machine::{Action, CommandHandler, Error, StateMachineApplyContext};
+use crate::{debug_if_leader, info_if_leader};
+use restate_invoker_api::InvokeInputJournal;
+use restate_service_protocol_v4::entry_codec::ServiceProtocolV4Codec;
+use restate_storage_api::invocation_status_table::{InvocationStatus, InvocationStatusTable};
+use restate_storage_api::journal_table_v2::JournalTable;
+use restate_types::identifiers::InvocationId;
+use restate_types::journal_v2::{
+    Command, CommandMetadata, CompletionId, EntryIndex, EntryMetadata, EntryType, NotificationId,
+    NotificationType,
+};
+use std::cmp;
+use tracing::{trace, warn};
+
+pub struct OnTrimCommand {
+    pub invocation_id: InvocationId,
+    pub invocation_status: InvocationStatus,
+    pub trim_point_command_entry_index: EntryIndex,
+}
+
+impl<'ctx, 's: 'ctx, S> CommandHandler<&'ctx mut StateMachineApplyContext<'s, S>> for OnTrimCommand
+where
+    S: JournalTable + InvocationStatusTable,
+{
+    async fn apply(self, ctx: &'ctx mut StateMachineApplyContext<'s, S>) -> Result<(), Error> {
+        let OnTrimCommand {
+            invocation_id,
+            invocation_status,
+            trim_point_command_entry_index,
+        } = self;
+
+        let is_invoked = matches!(invocation_status, InvocationStatus::Invoked(_));
+        let Some(mut in_flight_invocation_metadata) = invocation_status.into_invocation_metadata()
+        else {
+            info_if_leader!(
+                ctx.is_leader,
+                "Ignoring trim command because the invocation is not invoked nor suspended"
+            );
+            return Ok(());
+        };
+
+        // Validate the command first. The entry index must correspond to a command index.
+        if trim_point_command_entry_index == 0 {
+            info_if_leader!(
+                ctx.is_leader,
+                "Ignoring trim command because index is 0. You can't remove the input entry."
+            );
+            return Ok(());
+        }
+        let Some(trim_point_entry) = ctx
+            .storage
+            .get_journal_entry(invocation_id, trim_point_command_entry_index)
+            .await?
+        else {
+            info_if_leader!(
+                ctx.is_leader,
+                "Ignoring trim command because the given entry index doesn't exist"
+            );
+            return Ok(());
+        };
+        let Ok(trim_point_command) = trim_point_entry.decode::<ServiceProtocolV4Codec, Command>()
+        else {
+            info_if_leader!(
+                ctx.is_leader,
+                "Ignoring trim command because the given entry index doesn't correspond to a command entry"
```

Comment:
Is it possible to trim the an awakeable if it is the first thing that happens in a service handler?

```
const { id, promise } = ctx.awakeable<string>();
// send id to external service
await client.send(id);
const value = await promise;
```

Reply:
The awakeable completion/signal might arrive first (if this wasn't changed with the latest SDKs).

Reply:
no this is not possible.

Reply:
What prevents us from supporting this? I would expect that users will be surprised to learn about this limitation (probably at a point where it's already to late for them to take counter measures).

Reply:
What happens to the service state? Let's say I have a counter and I wanna time travel back to before the state was incremented. Looking at the code, this won't be possible. As a user this is probably surprising. Do we have somewhere written down the semantics of this feature and initial limitations? This would be helpful for me for the review and for users to understand what works and what not. Maybe you already have plans for how to evolve this feature as well. Cc @igalshilman

Reply:
> What happens to the service state? 

As described before, this command is expected to be alike killing an invocation in terms of guarantees, and similarly to killing we don't touch the state changes, keeping them as they are.

Reply:
> What prevents us from supporting this? I would expect that users will be surprised to learn about this limitation (probably at a point where it's already to late for them to take counter measures).

FYI I haven't ignore this, I'm still thinking about it. Perhaps I can alleviate the trim condition and allow trimming from a signal notification.
The reason I don't allow trimming from a completion is that it would be too confusing to understand: what happens to the related command? it restarts or not?! But this reasoning doesn't apply to signals, because they're externally triggered, so they don't have a relative command in the journal.

Reply:
Thanks for the clarification. I was thinking about two cases: a) Awakeables where the completion arrives before the awakeable command/promise and b) being able to trim everything up to input entry (basically kill and restart from the very beginning).

Reply:
Solved.